protobuf-provider/src/test/java/io/confluent/kafka/schemaregistry/protobuf/diff/SchemaDiffTest.java

```diff
@@ -0,0 +1,105 @@
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class SchemaDiffTest {
+
+  @Test
+  public void checkProtobufSchemaCompatibility() throws Exception {
+    ObjectMapper objectMapper = new ObjectMapper();
+    ArrayNode nodes = (ArrayNode) objectMapper.readTree(readFile("diff-schema-examples.json"));
+
+    for (int i = 0; i < nodes.size(); i++) {
+      final ObjectNode testCase = (ObjectNode) nodes.get(i);
+      String originalSchema = testCase.get("original_schema").asText();
+      String updateSchema = testCase.get("update_schema").asText();
+      ProtoFileElement original = ProtoParser.parse(Location.get("unknown"), originalSchema);
+      ProtoFileElement update = ProtoParser.parse(Location.get("unknown"), updateSchema);
+      final ArrayNode changes = (ArrayNode) testCase.get("changes");
+      boolean isCompatible = testCase.get("compatible").asBoolean();
+      final List<String> errorMessages = new ArrayList<>();
+      for (int j = 0; j < changes.size(); j++) {
+        errorMessages.add(changes.get(j).asText());
+      }
+      final String description = testCase.get("description").asText();
+
+      List<Difference> differences = SchemaDiff.compare(original, update);
+      final List<Difference> incompatibleDiffs = differences.stream()
+          .filter(diff -> !SchemaDiff.COMPATIBLE_CHANGES.contains(diff.getType()))
+          .collect(Collectors.toList());
+      assertThat(
+          description,
+          differences.stream()
+              .map(change -> change.getType().toString() + " " + change.getFullPath())
+              .collect(toList()),
+          is(errorMessages)
+      );
+      assertEquals(description, isCompatible, incompatibleDiffs.isEmpty());
+    }
+  }
+
+  @Test
+  public void checkCompatibilityUsingProtoFiles() throws Exception {
+
+    ResourceLoader resourceLoader = new ResourceLoader(
+        "/io/confluent/kafka/schemaregistry/protobuf/diff/");
+    ProtoFileElement original = resourceLoader.readObj("TestProto.proto");
+    ProtoFileElement update = resourceLoader.readObj("TestProto2.proto");
+
+    List<Difference> changes = SchemaDiff.compare(original, update);
+
+    assertTrue(changes.contains(new Difference(
+        Difference.Type.FIELD_NAME_CHANGED,
+        "#/TestMessage/2"
+    )));
+    assertTrue(changes.contains(new Difference(
+        Difference.Type.FIELD_SCALAR_KIND_CHANGED,
+        "#/TestMessage/2"
+    )));
+  }
+
+  public static String readFile(String fileName) {
+    ClassLoader classLoader = ClassLoader.getSystemClassLoader();
+    InputStream is = classLoader.getResourceAsStream(fileName);
+    if (is != null) {
+      BufferedReader reader = new BufferedReader(new InputStreamReader(is));
+      return reader.lines().collect(Collectors.joining(System.lineSeparator()));
+    }
+    return null;
+  }
```

Comment:
Nit: can we not reuse the ResourceLoader instead of this ?

Reply:
Ack